aten/src/ATen/ParallelOpenMP.h

```diff
@@ -30,15 +30,17 @@ inline void parallel_for(
   // returns false when omp_get_max_threads() == 1 inside nested "omp parallel"
   // See issue gh-32284
 
-#pragma omp parallel if (omp_get_max_threads() > 1 && !omp_in_parallel() && ((end - begin) > grain_size))
-  {
-    // choose number of tasks based on grain size and number of threads
-    // can't use num_threads clause due to bugs in GOMP's thread pool (See #32008)
-    int64_t num_threads = omp_get_num_threads();
-    if (grain_size > 0) {
-      num_threads = std::min(num_threads, divup((end - begin), grain_size));
-    }
+  // choose number of tasks based on grain size and number of threads
+  // If we are inside an OpenMP level, then we should only use one thread
+  // Otherwise, we can decide how many threads to assign any work to
+  // The remaining threads of the thread-pool won't have to do any wasteful work
+  int64_t num_threads = omp_get_level() ? 1 : omp_get_max_threads();
+  if (grain_size > 0) {
+    num_threads = std::min(num_threads, divup((end - begin), grain_size));
+  }
 
+#pragma omp parallel if (omp_get_max_threads() > 1 && !omp_in_parallel() && ((end - begin) > grain_size)) num_threads(num_threads)
```

Comment:
`!omp_in_parallel` check isn't required, as `num_threads` would be 1 if we're in an OpenMP level. So, `num_threads(1)` wouldn't result in any parallelization, as it is equivalent to an if clause being false. Reference: https://www.openmp.org/spec-html/5.0/openmpsu35.html